### PR TITLE
Implement relatively strong security and well architected setup for various parts

### PR DESCRIPTION
**Pull request by matthewzhaocc**
Implement security header standards for nunaserver

Implement strong security for nginx configuration according to latest web facing infra best practices

Implement a health check endpoint and clear dependency tracing for development docker compose

### CODE REVIEW

**Comment by matthewzhaocc** [comment]
Sorry, not dependency tracing, dependency startup order.

File: nunaserver/docker-compose.yml
Comment (clyde-johnston) [inline nunaserver/docker-compose.yml:36]: Does this change need to go into `deployment/docker-compose.prod.yml` too?  If so, will need to modify GHA to set the password on deployment.

Comment (matthewzhaocc) [reply nunaserver/docker-compose.yml:36]: Yep sorry did not notice it there, I'll add the code in a few hours to docker-compose.prod and .dev and take a look at adding to ci

Comment (matthewzhaocc) [reply nunaserver/docker-compose.yml:36]: Yeah reading through the github actions, I can add the csrf sed, but the token itself need to be added by a maintainer. A random 32char string should be perfect for it

Comment (clyde-johnston) [reply nunaserver/docker-compose.yml:36]: Yes, I have added a new secret called CSRF_SECRET_KEY.  Could add the following line to the `production.yml` GHA.

`sed -r -i 's/(CSRF_SECRET_KEY=).*$/\1${{ secrets.CSRF_SECRET_KEY }}/g' docker-compose.yml`

**Comment by clyde-johnston** [comment]
You might need to rebase your branch to `main` to pull in the port 80 redirection changes.

File: nunaserver/docker-compose.yml
Comment (matthewzhaocc) [reply nunaserver/docker-compose.yml:36]: yep its added with the latest commit

**Comment by matthewzhaocc** [comment]
after my sad attempt at rebase, should be good now

**Comment by matthewzhaocc** [comment]
oops, meant to put that commit in a different branch, but its already here, so removing it is way too hard ^ added flower monitoring to the local docker-compose, prod docker compose would require too much additional security configuration with oauth to be added easily.

**Comment by matthewzhaocc** [comment]
cors was improperly configured

File: deployment/docker-compose.dev.yml
Comment (bbworld1) [inline deployment/docker-compose.dev.yml]: Is this spelled wrong?

File: deployment/docker-compose.prod.yml
Comment (bbworld1) [inline deployment/docker-compose.prod.yml]: Here too; spelled incorrectly?

File: nunaserver/nunaserver/settings.py
Comment (bbworld1) [inline nunaserver/nunaserver/settings.py]: Correct the spelling here too.

File: deployment/docker-compose.prod.yml
Comment (bbworld1) [inline deployment/docker-compose.prod.yml:49]: `curl` is not part of the nunaweb-server image as of now. If we're using it, then we should add it to the Dockerfile build.

File: nunaserver/docker-compose.yml
Comment (bbworld1) [inline nunaserver/docker-compose.yml]: Should be `--address=0.0.0.0`, otherwise it will fail to recognize the address option.

File: nunaserver/nunaserver/server.py
Comment (bbworld1) [inline nunaserver/nunaserver/server.py:38]: Note that the addition of CSRF protection means the frontend needs to be modified to send the CSRF token with requests; otherwise, they won't go through. Either you'll need to add support for this in the frontend, or we will have to open an issue and add it internally.

**Review by bbworld1 — CHANGES_REQUESTED** [review]
Overall this looks pretty good besides a few minor issues. Main issue blocking this from being merged is that frontend needs to support sending CSRF headers in its requests.

**Comment by pavel-kirienko** [comment]
@matthewzhaocc are you planning to keep working on this or should we close it?

**Comment by matthewzhaocc** [comment]
Hey sorry ive been busy lately, ill circle around back to this soon

File: deployment/docker-compose.dev.yml
Comment (matthewzhaocc) [reply deployment/docker-compose.dev.yml]: Done

File: deployment/docker-compose.prod.yml
Comment (matthewzhaocc) [reply deployment/docker-compose.prod.yml]: Done

File: nunaserver/nunaserver/settings.py
Comment (matthewzhaocc) [reply nunaserver/nunaserver/settings.py]: Done

File: deployment/docker-compose.prod.yml
Comment (matthewzhaocc) [reply deployment/docker-compose.prod.yml:49]: curl is preinstalled

File: nunaserver/docker-compose.yml
Comment (matthewzhaocc) [reply nunaserver/docker-compose.yml]: Done

File: nunaserver/nunaserver/server.py
Comment (matthewzhaocc) [reply nunaserver/nunaserver/server.py:38]: No idea how frontend programming work, so ill file a issue

**Comment by matthewzhaocc** [comment]
Should be good besides the issue for frontend

**Comment by pavel-kirienko** [comment]
@bbworld1 can you please give it a look?

**Comment by bbworld1** [comment]
@matthewzhaocc Please rebase on latest main.

**Comment by pavel-kirienko** [comment]
Closing as stale.

**Pull request closed without merging**
